### ISSUE TITLE
Placeholders for root language for singleton cognate sets

### ISSUE THREAD

**Issue by PaulHeggarty** [lingdb/CoBL#246]
### Update Comments on First Version, 2016-08-10  
----
This is looking magnificent.  Just a few minor things to fix, cases I hadn't thought through, which I’ll add as check boxes at the top.  
  
- [x] When a cognate set has *both* a root ref language and a loan source language specified, the structure should be:  [root ref form] IN [root ref language] ≤ [form in source language] IN [source language]
- [x] When a cognate set has *only* a loan source language specified, the structure should be:   [form in source language] IN [source language], i.e. the form should precede the language name, and the language name would not need another ≤ in front of it.
- [x] Better than [space][comma][space] immediately after the alias letter would probably be [space]=[space], i.e. an equals sign surrounded by spaces.
- [x] Clearer than [Sgl Clade:] would be [Only in Clade:]
- [x] Clearer than [Sgl Lg:] would be [Only in Sgl Lg:]

If you can fix all these, then otherwise it looks ready to deploy, and will be a revelation.

The views in the other forms look great, especially the greying out of loanword cognate sets.  Please apply the same logic to them as listed above, though.  Could you also grey out the one that still looks brown, too, if it's not too much hassle?

----
### Original Description

- [ ] Add A filter box for the Lex. Count column in the `1-Meaning, All Cognate Sets` form, so that we can enter 1 to easily filter all cognate sets with a single member.  

**Where no data is entered in the `Root Ref. Language` field** … create placeholder text based on the following logic...

- [x] Where `Loan Event` is set to true (and `Loan Source Language` is not empty), then the placeholder for `Root Ref. Language` should be [≤ ] followed by the name of the `Loan Source Language`, and the `Root Ref. Form` placeholder should be [≤ ] followed by the `Form in Loan Source Language` entry.  In other words, the symbol ≤ is used to indicate 'loanword from'.  
- [x] Where Clade Count = 1 and Lex Count = 1, the placeholder for `Root Ref. Language` should be [Sgl Lg: ] followed by the name of the language in which that lexeme appears.  The `Root Ref. Form` placeholder should be the `Orthographic` entry for that lexeme.

That would complete stage 1 of this issue, already worth a deployment at that stage, please.  But also excellent would be a fuller implementation, not just to 'singleton languages' but to 'singleton clades', on all clade levels used in the Clades table.

- [x] Where Clade Count = 1 and Lex Count > 1, we need similar placeholders.  `Root Language` should be [Sgl Clade: ] followed by the name of the clade in which that lexeme appears.
This should use the same sort of logic as already used for working out which clade colour boxes to show.  But this should be taken further, and applied to all individual clades in our clades table, e.g. not just Celtic, 10-0, but taken down to levels Cl-1 as well, i.e. Brythonic 10-1 and Goidelic 10-2.  And even to level Cl-2 where appropriate:  as I said, all clades in our table.
That is, if the lexemes in a cognate class are found only within languages that are in 10-1, then the `Root Language` should be [Sgl Clade: Brythonic].  If they are found within languages that are in both 10-1 and 10-2, but not in any other clades at level Cl0 (i.e. not in 6-1 or something), then the `Root Language` should be [Sgl Clade: Celtic], the clade name for 10-0.

This issue may need more discussion for clarity!

### Original Entry for Discussion

I notice today 2016-07-28 that someone -- could you please make yourself known! -- has been doing a lot of work filling in blank cognate set root forms and root languages, e.g. here:
http://www.cobl.info/meaning/bark/
or here:
http://www.cobl.info/cognateclasslist/

In particular, the terms Indic, Iranic and Modern_Indic_Languages have started appearing here.  So too have lots of identifications of individual languages, presumably in cases where a lexeme is the only member of a cognate set, which is thus effectively unique to that language (for that meaning).
I would like to know who is doing this, because we are urgently working to a new system for specifying Root Languages, to be programmed next week. 

And whoever is doing this is not following correct policy -- note for example the root language 'Ossetic' for a form in Kashmiri.  There are various possibilities here (bad data, loanword not marked as a loanword event cognate set), but this should simply not be being done like this.  We need to consider making this an admin-only right, and soon.

Meanwhile, this also makes me think that to fill in the blanks far more efficiently but also accurately, and indeed for other good reasons, we should also have, please Jakob:
- A filter box for the Lex. Count column in the `1-Meaning, All Cognate Sets` form, so that we can easily filter all cognate sets with a single member.

In fact, it would be excellent if something like this could be implemented on `1-Meaning, All Languages` too, and perhaps some fast links to fill in the root language as the language for that record.  We've already discussed limiting the root language drop-down list on the basis of the clades that a cognate set is found in, so this is a similar thing.   

The main problem and scope for inconsistency here is in the vexed question of whether the root language is simply the earliest known one in which this root is found, or the earliest one in which it is found in specifically this meaning.  I think this is gong to be very hard to get consistency in, across many coders.  Part of the very important policy we need to draw up on this next week.

**Comment by m-scarborough** [comment]
I was the person recently working on 'bark'.  I was specifically trying to sort out classes E and N which I have specified as having earliest reconcilable antecedents in Middle Indic and Proto-Indo-Iranian respectively.  I admit to putting in the root forms there and specifying them as such.

I don't know where the things specified as Modern_Indic_Languages (or various other things with underscores that I've also noticed into the root languages form) came from.  Likewise where root languages are specified as things as 'Baltic', 'Waziri', 'Latvian', etc. are also not me.  I don't know where those came from.

For enabling consistency, a drop-down list may be the best solution here.  The problem is agreeing on which root language varieties need be included and what terminology to use for them.

**Comment by m-scarborough** [comment]
(Last post edited, to specify it was *yesterday* that I was working on 'bark'.  Further changes may have happened within the meaning since then.)

**Comment by PaulHeggarty** [comment]
No point in entering forms manually for any singleton cognate set, it can be done automatically, programmatically.  I’ll update this issues shortly with instructions for Jakob on what is needed.

**Comment by PaulHeggarty** [comment]
Putting together the following observations:
- There are many hundreds of cognate sets that contain just a single lexeme.
- A high proportion of these are caused by over-zealous entry of multiple synonyms, especially in Indic and Iranic.
- Many of these excessive synonyms will be eliminated over the next few weeks as those lists are revised, weeded out by adding lots of 'Not Swh' ticks, or even switched off as wholly unreliable. 
- These singleton cognate sets are uninformative as to Indo-European branch structure in any case, and only affect branch lengths.  Indeed as per the previous two points, they are generally bad news to the extent that many are excess synonyms.  They just exaggerate branch lenghts far beyond what they should be, the biggest specific problem in IELex that CoBL is setting out to fix.
- It is not worth spending lots of expert time on these, to fill in root form and root language, when they are so uninformative and irrelevant to all but 1 language in any one case.
- All we need to know is that these are singletons, and in which language they are.  
- It is especially not worth spending expert time to fill in root form and root language when -- because these are singletons with only one member lexeme from one language -- those values can be found and filled in automatically.

So I propose this solution:
* Nobody should manually fill  in these fields for singleton cognate sets.
* Rather, Jakob should programme things so that:
- in the condition where Lex. Count = 1, 
… the programme should …
- look up the single lexeme in that set
- look up the value in the `Orthographic` field for that lexeme, and fill it into the root field for that singleton cognate set
- look up the name of the language to which that lexeme belongs, and fill it into the root language field for that singleton cognate set, preceded by the text: "Singleton, in: "

One outstanding question is whether to make the root form and root language fields even uneditable, so long as Lex. Count = 1.  But that sounds like too much complexity for now.

One other question is whether to get Jakob to do this automatically, and programmatically fill in the root form and root language fields for all singleton cognate sets throughout the Jena 200 list, where those fields are currently blank.  I would vote yes.  So long as we used 'Singleton, in: " as a prefix, we should be able to reverse this operation if anything goes wrong, and tell apart manual entries from those computed automatically.

An alternative would be not to actually fill in these fields in the database, but only display them as dummy entries, recalculated each time.  This sounds like more complex programmatically, however.

**Comment by m-scarborough** [comment]
I'm still somewhat concerned about the idea of simply putting in the orthographic form as a the root form in singleton lexemes, from the point of data accuracy.  This may be a solution when there is no etymology, but for cases where there are good etymologies these should identified.  

A case in point within Hellenic is http://cobl.info/cognate/5312/, which has a very good Indo-European root etymology, but (at least prior to the entry of Cypriot and Italiot in the last month) would have otherwise been a singleton lexeme.

I understand why this is currently low-priority right now (I am now focusing on Cassandra's 2-branch list), but I think the singleton lexemes still need to be examined at some point—although definitely not until after the final lists with all not-swh ticks have been correctly assigned.

**Comment by PaulHeggarty** [comment]
Absolutely, where we know the etymology it should be put in as such.  The case you mention for Hellenic is just one of the ones that are problematic as I mentioned in an earlier comment:  great, old etymologies, but only found in that meaning in one of our languages…  
That said, this info is in fact irrelevant to the data export.  Indeed the point of the root form and root language is that without it, on the basic `1-Language, All Meanings` and `1-Meaning, All Languages` forms, and in the new rop-down box, one cannot tell what a cognate set 'is' from its alias letter alone.  

I see this as a critical inefficiency and danger for duplication that we must solve as soon as possible.

So on the contrary, I do not see it as low-priority at all.  But it is as very low-priority to be done _manually,_ wasting experts' time, that's all.  What we should do now is the automated solution, and then the experts come along and update in any cases such at the one you identify.  They will be many.  But the point of the 'singleton' automatically filled in data is a compromise for now to vastly clarify and improve cognacy coding.  That really does matter for the data export, and getting things for that ready and as reliable as possible in time for 20th August.  

The deeper etymologies can be done later, they don’t actually go into the data export.  They are only crucial, of course, whenever people have not realised that a singleton is actually part of an existing cognate set!

**Comment by CasFre** [comment]
I basically support the automatic solution for the reasons that @PaulHeggarty has named.
The entries with underscores and the manually put in individual language names put in as root languges that you noticed indeed stem from some months ago when there were even more blank fields and we did not have the colour boxes yet. So these were basically temporary entries in order to provide some overview to me (and other people who might do new entries and want to add them to an existing cognate set) so that one knew what was in every cognate set.

**Comment by cormacanderson** [comment]
I agree with @PaulHeggarty that we should attempt to automatise this, but think that the reservations of @m-scarborough have to be taken into consideration. The cognate sets for DO whose Lex. Count = 1 are a beautiful illustration of this:
![screen shot 2016-08-01 at 19 56 35](https://cloud.githubusercontent.com/assets/8975419/17303722/2db1a24e-5822-11e6-9a48-3261a5b40d30.png)
Four of the six have PIE etymologies (I presume http://cobl.info/cognate/5936/ should be merged into http://cobl.info/cognate/5037/). For this reason, I would suggest a slightly less drastic semi-automated solution.
One option would be to turn the "Merge" checkboxes on the `1 Mng All Cog Sets` form into "Select" checkboxes, as we have on the `1 Mng All Lgs` html form, and then put a "Merge" button and an "Assign Unique Root" button at the top of the page. The latter would automatically look up the lexeme and language in each cognate class selected and automatically copy these over as `Root Form` and `Root Language`. We would need an error message if any language was selected for which Lex. Count ≠ 1.
A drawback of this is that multiple clicks are necessary to establish the content of a cognate set with only one member on the `1 Mng All Cog Sets` html form. For this reason, an alternative would be to enable this same functionality (i.e. with a button) on the `1Mng All Lgs` html form, where the Select boxes are already in place and it is relatively easy to see the cognate sets with only one member (white vs colour when you sort by cognate alias). This would be especially efficient if we could filter so that we would see e.g. only reference languages for cognate sets with multiple members.
These solutions would semiautomate the process, speeding things up drastically, but giving the experts the chance to filter out the lexemes which might have good older etymologies.

**Comment by m-scarborough** [comment]
> (I presume http://cobl.info/cognate/5936/ should be merged into http://cobl.info/cognate/5037/).

Well spotted.  I've just merged those.

**Comment by runjak** [comment]
When using the current placeholder code I've written for the cognate class selection I get something like this:

![screenshot_2016-08-10_12-27-10](https://cloud.githubusercontent.com/assets/5417642/17550564/d4a87dc8-5ef5-11e6-97ef-32b12b7a9719.png)

I'll also add it as a placeholder to the `Root ref. form` and `Root ref. language` inputs before deploying it.

**Comment by PaulHeggarty** [comment]
This is looking magnificent.  Just a few minor things to fix, cases I hadn't thought through, which I’ll add as check boxes at the top.  
  
- [ ] When a cognate set has *both* a root ref language and a loan source language specified, the structure should be:  [root ref form] IN [root ref language] ≤ [form in source language] IN [source language]
- [ ] When a cognate set has *only* a loan source language specified, the structure should be:   [form in source language] IN [source language], i.e. the form should precede the language name, and the language name would not need another ≤ in front of it.
- [ ] Better than [space][comma][space] immediately after the alias letter would probably be [space]=[space], i.e. an equals sign surrounded by spaces.
- [ ] Clearer than [Sgl Clade:] would be [Only in Clade:]
- [ ] Clearer than [Sgl Lg:] would be [Only in Sgl Lg:]

If you can fix all these, then otherwise it looks ready to deploy, and will be a revelation.

**Comment by runjak** [comment]
Alright - I'll look after your additions in a moment.
Current placeholder code in the rest of the site looks like this:

1 meaning all languages view:
![screenshot_2016-08-10_12-44-15](https://cloud.githubusercontent.com/assets/5417642/17551088/66f215c0-5ef8-11e6-8fc9-eccd1e10be18.png)
1 language all meanings view:
![screenshot_2016-08-10_12-45-30](https://cloud.githubusercontent.com/assets/5417642/17551089/66f27506-5ef8-11e6-9328-4dc1bb7b24ce.png)
1 meaning all cognate sets view:
![screenshot_2016-08-10_12-46-18](https://cloud.githubusercontent.com/assets/5417642/17551101/74fdd51e-5ef8-11e6-8594-7a8a40872db7.png)

**Comment by runjak** [comment]
I have a bit of trouble with the current topmost checkbox in this issue:

> When a cognate set has both a root ref language and a loan source language specified, the structure should be: [root ref form] IN [root ref language] ≤ [form in source language] IN [source language]

Currently I'm computing the placeholders for `Root ref. form` and `Root ref. language` separately and I would like to keep it that way if possible. This is because of two things:

1. Placeholders can also be used for table inputs.
2. The code calculating the placeholders would need more reorganizing if we want a combined thing in one place while not combined in another.

A string in the form of
```
[root ref form] IN [root ref language] ≤ [form in source language] IN [source language]
```
would not satisfy this because it mixes both things.

**Comment by PaulHeggarty** [comment]
I don’t get this, though, because some of your existing entries have the names of two languages, both loan source language and root ref language, I think- but maybe I’ve got that wrong.  If that is not so, then work out a solution with Cormac, please.

**Comment by runjak** [comment]
It's just that it will take additional work because currently I have a nice solution that works for both things independently (see [models.py:930-965](https://github.com/lingdb/CoBL/blob/855071250e7c5bc92129a4f5074529919768a0d5/ielex/lexicon/models.py#L930-L965)). So now I'll have to add a third thing that reimplements parts of that code. :persevere:

**Comment by cormacanderson** [comment]
If this is really impossible, or would take an inordinate amount of time, I would suggest an alternative string of the form `[root ref form] IN [root ref language] ≤ [source language]`. That keeps the `Root ref. form` and `Root ref. language` separate. Would that suit better?

**Comment by runjak** [comment]
No, it's not impossible.
It just complicates things more than I've hoped and inferred originally.
Basically I'll just have to fortify a bit more and get this over with.

**Comment by cormacanderson** [comment]
Also, I'm not sure I understand either. You currently have strings of the form `*pisk ≤ Latin = Protto-Brythonic ≤ piscis. Surely this is mixing the two things as well?

**Comment by cormacanderson** [comment]
That ordering, by the way, is unclear, but if you explain the problem a little better, perhaps we can find a solution than means you don't have a lot more work to do, and which is yields forms which are more intuitive for us.

**Comment by runjak** [comment]
Yes, but it only mixes them client side.
Server side I only have strings `Latin ≤ Proto-Brythonic` and `*pisk ≤ piscis` at the moment.
The client is not doing more than gluing these parts together with ` = ` at the moment.
To get strings that have even a form of `*pisk ≤ Latin` I need to consider root form and root language at once which is currently not happening.

**Comment by cormacanderson** [comment]
Okay, can you do `[Root ref. form] IN [Root ref. lg.] ≤ [Loan source lg.] = [Loan source form]? or is that equally a problem.

**Comment by cormacanderson** [comment]
There is currently also a bug. In the `1 Mng, all Cog sets` table, the placeholders are mixed up if there `loan source lg.` and `loan source form` are entered. See below:
![screen shot 2016-08-10 at 14 06 32](https://cloud.githubusercontent.com/assets/8975419/17553124/5177d02a-5f04-11e6-9182-1ee93fceedb0.png)
That bottom entry is wrong. The `Loan source form` is in the `Root language` field, while the `Loan source lg` is in the `Root form` field.

**Comment by cormacanderson** [comment]
As the above is a singleton language, it should, I think, be marked the same way as other singleton languages, i.e. with the placeholders being the orthographic form and "Only in Sgl Lg.: Lg x".

**Comment by runjak** [comment]
Alright - sorry for the delay.
I've got something like this now:
![screenshot_2016-08-11_10-11-11](https://cloud.githubusercontent.com/assets/5417642/17582435/f2253cae-5fab-11e6-8ac6-10f200508b67.png)

I'm not sure about the `Proto-Indo-European ≤ Baltic` parts, but I feel like it's getting closer to what we want here.

**Comment by cormacanderson** [comment]
Yes, that looks more like it, although I hope those are not real examples, or we have problems. Of the above, D shows the behaviour we are looking for, while H and K show those in which the `Loan source form` is unspecified, right? I presume  A and F are outputs for the unusual cases in which the `Root ref. form` is present but the `Root ref. Lg.` absent, while the `Loan source form` is absent and the `Loan source Lg.` present. Is this correct?

**Comment by runjak** [comment]
Currently it should work like this:
* For the part before the `≤`:
  * If we have a `Root ref. form` and a `Root ref. Language`:
    * Put `Root ref. form IN Root ref. Language`
  * If we only have `Root ref. form` or `Root ref. Language`:
    * Put the one we have
  * If we have neither `Root ref. form` nor `Root ref. Language`:
    * Don't even put the `≤`
* For the part after the `≤`:
  * If we have a `root form placeholder` and a `root language placeholder`:
    * Put `root form placeholder IN root language placeholder`
  * If we only have a `root form placeholder` or a `root language placeholder`:
    * Put the one we have
  * If we have neither `root form placeholder` nor `root language placeholder`:
    * Don't even put the `≤`

Usually the placeholders are computed like this:
* `root form placeholder`:
  * If we have a `Loan source form` we use it
  * If no `Loan source form` but only `1` lexeme in cognate set:
    * Use `source form` of that single lexeme
  * Otherwise give `''`
* `root language placeholder`:
  * If we have a `Loan source lg` we use it
  * If no `Loan source lg` but only `1` lexeme in cognate set:
    * Use `language name` for that single lexeme
  * Search for the most specific clade spanning all languages in the cognate set:
    * If such a clade is found use it

Now one thing I'm not sure about is this:
Should the assignments be restricted only to the root and loan data or is the above strategy OK?

**Comment by cormacanderson** [comment]
Okay, now I see what is happening. I think there is potential for confusion here.  
We want a quite fixed syntax for these expressions, or else it will get confusing. We also want to be able to differentiate a placeholder from an entered root (except when the set has only one lexeme, in which case it does not matter). The fixed syntax should be of the form `Root ref. form IN Root ref. lg. ≤ Loan source form IN Loan source lg.` and we enter what we can of that. If either `Root ref. form or `Root ref. lg.` are missing, then we need a placeholder. There should be a `≤` only if the `Cognate class` is marked [+Loan Event], i.e. we use that sign to mark a borrowing. 

* For the part before the `≤` (which should always be present):
  * If we have a `Root ref. form` and a `Root ref. Language`:
    * Put `Root ref. form IN Root ref. Language`
  * If we only have `Root ref. form` or `Root ref. Language`:
    * Put the one we have  
  * If either are missing, calculate placeholders
  
* For the part after the `≤` (which should be present only if the set is marked [+Loan event]:
  * If we have a `Loan source form` and a `Loan source lg.`:
    * Put `Loan source form IN Loan source lg.`
  * If we only have a `Loan source form` or a `Loan source lg.`:
    * Put the one we have
  * Otherwise, leave blank, but keep the `≤` to show that the set derives from a loanword event.
  
This yields the possibilities for sets which are marked [+Loan Event] and have `Root ref. form` and `Root ref. Lg.` entered:  
`Root ref. form IN Root ref. lg. ≤ Loan source form IN Loan source lg.`
`Root ref. form IN Root ref. lg. ≤ Loan source form` (no `Loan source lg` entered)
`Root ref. form IN Root ref. lg. ≤ Loan source lg.` (no `Loan source form` entered)
`Root ref. form IN Root ref. lg. ≤` (neither `Loan source form` nor `Loan source lg.` entered)
  
If either `Root ref. form` or `Root ref. Lg` are missing, they should be computed:
* `root form placeholder`:
  * If only `1` lexeme in cognate set:
    * Use `source form` of that single lexeme
  * If no. of lexemes in set is > 1  and we have a `Loan source form` we use it in brackets, i.e. `(Loan source form)` (this way we can identify it as a placeholder, not an entered root)
 ( * If no. of lexemes in set is > 1 and we do not have `Loan source lg.`, but the `root language placholder` is a clade name, and there is a lexeme from a reference language in the set then use this form, in brackets (forget this bit if it is too difficult to compute and if you do do it, be careful that some clades have more than one root lg. - you can pick the one with the lowest lex. ID, for example) )
  * Otherwise give `''`
* `root language placeholder`:
  * If only `1` lexeme in cognate set:
    * Use `language name` for that single lexeme
  * If no. of lexemes in cognate set is >1, search for the most specific clade spanning all languages in the cognate set:
    * If such a clade is found use it, in brackets
    * If no such clade is found but we have a `Loan source lg.` use it in brackets, i.e. `(Loan source lg.)`
  * Otherwise give `''`  

Does that sound doable in a reasonably short period of time? There are also other pressing priorities, as usual!

**Comment by runjak** [comment]
This sounds good and I think the issue is clearer to me now!
Thanks for going trough the whole decision tree here :)

**Comment by runjak** [comment]
My current version looks like this:
![screenshot_2016-08-11_11-51-56](https://cloud.githubusercontent.com/assets/5417642/17585081/0c37a20e-5fba-11e6-9dc4-b9ef7cd2c9c5.png)

It's a bit unfortunate how the selection box breaks out of the modal because it's too wide.
But this is probably something I can attend to later.

**Comment by runjak** [comment]
Second screenshot for further illustration:
![screenshot_2016-08-11_12-06-29](https://cloud.githubusercontent.com/assets/5417642/17585445/154ba4a6-5fbc-11e6-9a56-c4dd9859dea9.png)

**Comment by cormacanderson** [comment]
We don't need the bits in brackets about the `Loan source form` or `Loan source lg.` as this can be inferred. Otherwise it looks good

**Comment by cormacanderson** [comment]
So, for W above, just `form IN Nepali ≤` tells us already that it's a loan, but from what has not been entered.

**Comment by runjak** [comment]
Sweet. I'll fix that and let this issue be closed on deployment.

**Comment by CasFre** [comment]
Hi guys,
It's not a singleton cognate set but there seems to be a mistake with the placeholders in root ref. form and root ref. language in a loanword cognate set, namely 2277, so I put it here:
![root ref form_loanword cognate sets](https://cloud.githubusercontent.com/assets/17431124/17866193/e4a8988e-68a5-11e6-8764-9bb1c17f9399.png)

**Comment by cormacanderson** [comment]
This is as it should be. Seeing as no root reference form is entered, and there is more than one language in the set, the loan source language and loan source form (in brackets) are the placeholders.

**Comment by CasFre** [comment]
Okay, sorry then.
It seems reasonable to double the information in that case, although it might be helpful to distinguish it from other cases where it's not a loan source language... but maybe we can worry about that after 26nd...

**Comment by cormacanderson** [comment]
No problem, it might not be the best policy that, but in lieu of an obvious alternative...
